### PR TITLE
Refactor FairSchedulableBuilder

### PR DESCRIPTION
**Pull request by jerryshao**
1. FairScheduler configuration can be read from class path if it is not set explicitly.
2. Add missing file handler close.
3. Simplify structure.

Thanks
Jerry

### CODE REVIEW

**Comment by jey** [comment]
Jenkins, please test this.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/SchedulableBuilder.scala [hunk 1]
```diff
@@ -64,48 +65,26 @@ private[spark] class FairSchedulableBuilder(val rootPool: Pool)
...
+          new FileInputStream(f)
+        } getOrElse {
+          getClass.getClassLoader.getResourceAsStream(DEFAULT_SCHEDULER_FILE)
         }
-      } else {
-        throw new java.io.FileNotFoundException(
-          "Fair scheduler allocation file not found: " + schedulerAllocFile)
       }
+
+      is foreach { i => buildFairSchedulerPool(i) }
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/SchedulableBuilder.scala]: Don't use infix style. Use is.foreach ...

File: core/src/main/scala/org/apache/spark/scheduler/cluster/SchedulableBuilder.scala [hunk 2]
```diff
@@ -64,48 +65,26 @@ private[spark] class FairSchedulableBuilder(val rootPool: Pool)
...
+    } finally {
+      is.foreach(_.close)
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/SchedulableBuilder.scala]: close has side-effect, so leave the parenthesis there

```
_.close()
```

File: core/src/main/scala/org/apache/spark/scheduler/cluster/SchedulableBuilder.scala [hunk 3]
```diff
@@ -115,6 +94,41 @@ private[spark] class FairSchedulableBuilder(val rootPool: Pool)
...
+      var schedulingMode = DEFAULT_SCHEDULING_MODE
+      var minShare = DEFAULT_MINIMUM_SHARE
+      var weight = DEFAULT_WEIGHT
+
+      val xmlSchedulingMode = (poolNode \ SCHEDULING_MODE_PROPERTY).text
+      if (xmlSchedulingMode != "") {
+        try {
+          schedulingMode = SchedulingMode.withName(xmlSchedulingMode)
+        } catch {
+          case e: Exception => logInfo("Error xml schedulingMode, using default schedulingMode")
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/SchedulableBuilder.scala]: Change this to logWarning (technically not your problem but let's do it in the same pull request)

File: core/src/main/scala/org/apache/spark/scheduler/cluster/SchedulableBuilder.scala [hunk 4]
```diff
@@ -64,48 +65,26 @@ private[spark] class FairSchedulableBuilder(val rootPool: Pool)
...
-          }
-
-          val pool = new Pool(poolName, schedulingMode, minShare, weight)
-          rootPool.addSchedulable(pool)
-          logInfo("Created pool %s, schedulingMode: %s, minShare: %d, weight: %d".format(
-            poolName, schedulingMode, minShare, weight))
+    var is: Option[InputStream] = None
+    try {
+      is = Option {
+        schedulerAllocFile map { f =>
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/SchedulableBuilder.scala]: No infix style. See below

**Comment by jey** [comment]
Jenkins, this is ok to test.

**Comment by jerryshao** [comment]
Hi @rxin thanks for your reviewing, I've updated the code according to your comments.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/SchedulableBuilder.scala [hunk 5]
```diff
@@ -64,48 +63,26 @@ private[spark] class FairSchedulableBuilder(val rootPool: Pool)
...
-            weight = xmlWeight.toInt
-          }
-
-          val pool = new Pool(poolName, schedulingMode, minShare, weight)
-          rootPool.addSchedulable(pool)
-          logInfo("Created pool %s, schedulingMode: %s, minShare: %d, weight: %d".format(
-            poolName, schedulingMode, minShare, weight))
+    var is: Option[InputStream] = None
+    try {
+      is = Option{
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/SchedulableBuilder.scala]: Add a space before all "{". Also for the map, getOrElse, foreach below

File: core/src/main/scala/org/apache/spark/scheduler/cluster/SchedulableBuilder.scala [hunk 6]
```diff
@@ -115,6 +92,41 @@ private[spark] class FairSchedulableBuilder(val rootPool: Pool)
...
+      var schedulingMode = DEFAULT_SCHEDULING_MODE
+      var minShare = DEFAULT_MINIMUM_SHARE
+      var weight = DEFAULT_WEIGHT
+
+      val xmlSchedulingMode = (poolNode \ SCHEDULING_MODE_PROPERTY).text
+      if (xmlSchedulingMode != "") {
+        try {
+          schedulingMode = SchedulingMode.withName(xmlSchedulingMode)
+        } catch {
+          case e: Exception => logWarning("Error xml schedulingMode, using default schedulingMode")
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/SchedulableBuilder.scala]: Catch only NoSuchElementException here.

**Comment by rxin** [comment]
Thanks, Jerry. Suggested a couple other changes. Otherwise this looks good and can be merged once you fix those couple minor things.

**Comment by jerryshao** [comment]
Hi @rxin, I've changed the code according to your comments, thanks a lot.

**Comment by rxin** [comment]
Jenkins, test this please.

**Comment by rxin** [comment]
Thanks. We also need to merge this back into branch-0.8 after 0.8.0 is released.

**Comment by rxin** [comment]
I've merged this into asf spark branch-0.8.

**Pull request merged**